src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReader.java

```diff
@@ -0,0 +1,304 @@
+  private String formattedTableName;
+  private String tableNameField;
+  private List<? extends ETApiObject> results;
+  private Iterator<? extends ETApiObject> iterator;
+  private ETApiObject row;
+
+  MarketingCloudRecordReader(MarketingCloudSourceConfig pluginConf) {
+    this.pluginConf = pluginConf;
+  }
+
+  @Override
+  public void initialize(InputSplit split, TaskAttemptContext context) {
+    this.split = (MarketingCloudInputSplit) split;
+    this.pos = 0;
+  }
+
+  @Override
+  public boolean nextKeyValue() throws IOException {
+    try {
+      if (results == null) {
+        fetchData();
+      }
+
+      if (!iterator.hasNext()) {
+        return false;
+      }
+
+      row = iterator.next();
+
+      pos++;
+    } catch (Exception e) {
+      LOG.error("Error in nextKeyValue", e);
+      throw new IOException("Exception in nextKeyValue", e);
+    }
+    return true;
+  }
+
+  @Override
+  public NullWritable getCurrentKey() {
+    return NullWritable.get();
+  }
+
+  @Override
+  public StructuredRecord getCurrentValue() throws IOException {
+    StructuredRecord.Builder recordBuilder = StructuredRecord.builder(schema);
+
+    if (pluginConf.getQueryMode() == SourceQueryMode.MULTI_OBJECT) {
+      recordBuilder.set(tableNameField, formattedTableName);
+    }
+
+    try {
+
+      convertRecord(recordBuilder, row);
+
+    } catch (Exception e) {
+      LOG.error(String.format("Error decoding row from table %s", tableName), e);
+      throw new IOException(String.format("Error decoding row from table %s", tableName), e);
+    }
+    return recordBuilder.build();
+  }
+
+  @Override
+  public float getProgress() throws IOException, InterruptedException {
+    return pos / (float) split.getLength();
+  }
+
+  @Override
+  public void close() throws IOException {
+  }
+
+  private void fetchData() {
+    object = SourceObject.valueOf(split.getObjectName());
+    tableName = split.getTableName();
+    formattedTableName = tableName.replaceAll("-", "_");
+
+    if (object == SourceObject.DATA_EXTENSION) {
+      dataExtensionKey = tableName.replaceAll(MarketingCloudConstants.DATA_EXTENSION_PREFIX, "");
+    }
+    tableNameField = pluginConf.getTableNameField();
+
+    //Get the table data
+    try {
+      MarketingCloudClient client = MarketingCloudClient.create(pluginConf.getClientId(), pluginConf.getClientSecret(),
+                                                                pluginConf.getAuthEndpoint(),
+                                                                pluginConf.getSoapEndpoint());
+
+      //Fetch data
+      if (object == SourceObject.DATA_EXTENSION) {
+        results = client.fetchDataExtensionRecords(dataExtensionKey, object.getFilter());
+      } else {
+        results = client.fetchObjectRecords(object);
+      }
+
+      LOG.debug("size={}", results.size());
+      if (!results.isEmpty()) {
+        fetchSchema(client);
+      }
+    } catch (Exception e) {
+      results = Collections.emptyList();
```

Comment:
How should user be reacting to this error message.

Reply:
Resolved.